virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1251 @@
+	printf(" -n, --netns-path   Network namespace path (required)\n");
+	printf(" -p, --pod-id       Pod ID (required)\n");
+	printf(" -r, --runtime-path Runtime path (required)\n");
+	printf(" -v, --version      Show version\n");
+	printf("\n");
+}
+
+/*
+ * Print version information.
+ */
+void print_version(void)
+{
+	printf("%s v0.1\n", PROGRAM_NAME);
+}
+
+/*
+ * Free internal fields of the watcher_params structure.
+ */
+void free_watcher_params(struct watcher_params *params)
+{
+	if (!params) {
+		return;
+	}
+
+	free(params->netns_path);
+	params->netns_path = NULL;
+
+	free(params->pod_id);
+	params->pod_id = NULL;
+
+	free(params->runtime_path);
+	params->runtime_path = NULL;
+}
+
+/*
+ * Free internal fields of the ip_addr structure.
+ */
+void free_ip_addr(struct ip_addr *ipa) {
+	struct ip_addr *ipa_next, *ipa_curr;
+
+	if (!ipa) {
+		return;
+	}
+
+	free(ipa->addr);
+	ipa->addr = NULL;
+
+	for (ipa_curr = ipa->next; ipa_curr; ipa_curr = ipa_next) {
+		ipa_next = ipa_curr->next;
+
+		free(ipa_curr->addr);
+		ipa_curr->addr = NULL;
+
+		free(ipa_curr);
+		ipa_curr = NULL;
+	}
+}
+
+/*
+ * Free internal fields of the iface structure.
+ */
+void free_iface(struct iface *nif) {
+	int i;
+
+	if (!nif) {
+		return;
+	}
+
+	free(nif->hw_addr);
+	nif->hw_addr = NULL;
+
+	free(nif->name);
+	nif->name = NULL;
+
+	free(nif->mtu);
+	nif->mtu = NULL;
+
+	if (nif->ip_addrs) {
+		free_ip_addr(nif->ip_addrs);
+	}
+	free(nif->ip_addrs);
+	nif->ip_addrs = NULL;
+}
+
+/*
+ * Free internal fields of the route structure.
+ */
+void free_route(struct route *rt) {
+	free(rt->src);
+	rt->src = NULL;
+
+	free(rt->dst);
+	rt->dst = NULL;
+
+	free(rt->gw);
+	rt->gw = NULL;
+
+	free(rt->dev);
+	rt->dev = NULL;
+}
+
+/*
+ * Modify the global list of interfaces by inserting a new ip_addr structure
+ * into the linked list of the specified interface.
+ */
+void insert_ip_addr(int idx, char addr[], unsigned char family)
+{
+	struct ip_addr *ipa = (struct ip_addr *)
+		malloc(sizeof(struct ip_addr));
+
+	ipa->addr = strdup(addr);
+	ipa->family = family;
+	ipa->next = NULL;
+
+	if (!iface_list[idx].ip_addrs) {
+		iface_list[idx].ip_addrs = ipa;
+		return;
+	}
+
+	struct ip_addr *tmp = iface_list[idx].ip_addrs;
+	while (tmp->next) {
+		tmp = tmp->next;
+	}
+
+	tmp->next = ipa;
+}
+
+/*
+ * Modify the global list of interfaces by removing an existing ip_addr
+ * structure from the linked list of the specified interface.
+ */
+void delete_ip_addr(int idx, char addr[])
+{
+	bool found = false;
+
+	if (!iface_list[idx].ip_addrs) {
+		return;
+	}
+
+	struct ip_addr *tmp = iface_list[idx].ip_addrs;
+	if (tmp->addr && !strcmp(tmp->addr, addr)) {
+		iface_list[idx].ip_addrs = iface_list[idx].ip_addrs->next;
+		goto free_node;
+	}
+
+	struct ip_addr *prev = tmp;
+	while (tmp->next) {
+		prev = tmp;
+		tmp = tmp->next;
+
+		if (tmp->addr && !strcmp(tmp->addr, addr)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		printf("No IP address matching %s for interface %s",
```

Comment:
This seems to be a warning so I'd print to `stderr` and prefix with `WARNING:`.

Reply:
I have replaced all warnings and errors with `fprintf(stderr...`